pandas/core/algorithms.py

```diff
@@ -445,19 +446,27 @@ def _factorize_array(values, check_nulls, na_sentinel=-1, size_hint=None):
     values : ndarray
     check_nulls : bool
         Whether to check for nulls in the hashtable's 'get_labels' method.
+        Nulls are always checked when `na_value` is specified.
```

Comment:
if you are going to add this kwarg (which is ok), then remove ``check_nulls``  which is redundant (otherwise this could be ``None``), and would simply default ``na_value`` to ``-1``

Reply:
Thought I posted a comment about that, but I guess I deleted it.

AFAICT `check_nulls` is only useful for `DatetimeIndex`, so that these two are different?

```python
In [2]: pd.factorize([2**-63, 0])
Out[2]: (array([0, 1]), array([1.08420217e-19, 0.00000000e+00]))

In [3]: pd.factorize(pd.DatetimeIndex([None, 0]))
Out[3]:
(array([-1,  0]),
 DatetimeIndex(['1970-01-01'], dtype='datetime64[ns]', freq=None))
```

Once DatetimeIndex is an ExtensionArray, `pd.factorize` will dispatch to `DatetimeIndex.factorize`, which can call `_factorize_array(..., na_value=iNaT)`. I think I can do this now, it just makes the logic a bit ugly. Will give it a shot.

Reply:
Ahh, may not be able to remove `check_nulls` entirely. I needed it for factorizing boolean arrays. Anyway, this is what I had in mind:

```diff
diff --git a/pandas/core/algorithms.py b/pandas/core/algorithms.py
index bf192cdb2..21ec17054 100644
--- a/pandas/core/algorithms.py
+++ b/pandas/core/algorithms.py
@@ -511,7 +511,7 @@ def factorize(values, sort=False, order=None, na_sentinel=-1, size_hint=None):
     values = _ensure_arraylike(values)
     original = values
 
-    if is_categorical_dtype(values):
+    if is_categorical_dtype(values) or is_datetime64_any_dtype(values):
         values = getattr(values, '_values', values)
         labels, uniques = values.factorize()
         dtype = original.dtype
diff --git a/pandas/core/indexes/datetimelike.py b/pandas/core/indexes/datetimelike.py
index b906ea0f4..bd33828f2 100644
--- a/pandas/core/indexes/datetimelike.py
+++ b/pandas/core/indexes/datetimelike.py
@@ -984,6 +984,13 @@ class DatetimeIndexOpsMixin(object):
 
         return algorithms.isin(self.asi8, values.asi8)
 
+    def factorize(self, sort=False, na_sentinel=-1):
+        from pandas.core.algorithms import _factorize_array
+        l, u = _factorize_array(self, True, na_sentinel=na_sentinel,
+                                na_value=iNaT)
+        u = self._constructor(u)
+        return l, u
+
     def shift(self, n, freq=None):
         """
         Specialized shift which produces a DatetimeIndex
```

Will have to fix it to check for timedeltaindex too, handle sorting, but not too bad. Worth doing here, or a separate PR?

Reply:
Got it. Will be able to remove `check_nulls` entirely without too much effort.